### DISCUSSION TITLE
How to Optimize Incremental Deletion Scenarios

### DISCUSSION THREAD

**Discussion by qi49125** [milvus-io/milvus#19048] [Ideas & Feature requests]
RT

**Comment by qi49125** [comment]
[Bug]: load collection failed，then auto release from memory #18860
![image](https://user-images.githubusercontent.com/32016286/188588934-1a0b3899-3191-4490-817b-4a3cac59e0d4.png)

**Comment by qi49125** [comment]
Background: In 2.1.0, after a certain amount of data is deleted from the collection, the collection cannot be loaded to the memory and searched for. Therefore, it is suspected that the compact problem is caused.

**Comment by jiaoew1991** [comment]
I will try to explain current logic of deletion

**Reply by jiaoew1991** [reply]
To understand today's storage design, you need to understand several limitations of vector storage
1. Before querying the data, it needs to be loaded into memory and perform large-scale MPP calculations to meet the needs of millisecond latency
2. Indexes **do not support updates**
3. Indexing is a **resource-intensive and time-consuming** operation
4. The amount of data is very small, and the indexing effect is very poor

In order to solve the above problems and have good extensibility, milvus has designed the following markup deletion mechanism

First, shard and segment are divided into shards and segments at the storage level, and the design is detailed in [data-insertion-and-data-persistence](https://milvus.io/blog/deep-dive-4-data-insertion-and-data-persistence.md)
The data of insert and delete will be written to the DML log queue, and the datanode node will forward the delete record in the DML to the delta queue, and each querynode that loads the data will subscribe to the delta queue, delete the data and the corresponding segment binding, and generate a bitset according to the filter criteria in the query scan for query results filtering

Delet records will be merged with a segment through a scheduled `compaction` to complete the true deletion

**Reply by jiaoew1991** [reply]
The current design has the following limitations:
1. When there are many segments, the read amplification of the Delta queue is very serious, and the resources are wasted
2. Wasted topic resources, `pulsar` as a log broker is fine, `kafka`'s topic resources have  limitations
3. When used with the time travel function, it can be deleted after a long time, and the memory is wasted

**Reply by qi49125** [reply]
According to your original design, it should be logically no problem to delete data. Why is it abnormal to delete more than 10% of the data set now, and it is impossible to load memory and query?
In other words, if before deleting the data, delete the index first, then delete and merge the data, and finally recreate the index, can this problem be avoided?

**Reply by jiaoew1991** [reply]
First of all, because there is a deletion record in the log queue, the `querynode` needs to subscribe to the corresponding shard topic after loading the segment, which leads to if a shard segment is loaded by N `querynode`, the subscription will be magnified N times, and the larger the proportion of deleted data, the more records need to be read, especially in the scenario of `querynode` restart and new `querynode` online, you need to consume the deletion record from the begininig. This causes the `querynode` to start very slowly.

In addition, the index cannot be deleted when the data is deleted, which will cause the query to fall back to the scene of brute force search, which will lead to a sharp decline in query performance, and the querynode will not load the original vector into memory in the case of the index is built, if the index is deleted, it is also necessary to load the original vector into memory, which will cause more pressure on the memory
